### ISSUE TITLE
Dependency substitution rules

### ISSUE THREAD

**Issue by bigdaz** [gradle/gradle#3880]
Presently, dependency substitution rules do not allow specifying several things that are available in a dependency declaration, such as `classifier`, `extension`, `configuration` and `artifact`. There are valid use cases for these things, and we may extend the existing rules support, or find alternative solutions using component metadata rules and dependency constraints.

**Comment by TWiStErRob** [comment]
Example use case for `classifier` + `extension`:
```gradle
// normal dependency
testImplementation "foo:bar:1.0:test-helpers@jar"

// but sometimes (based on flag) I want to build them locally for development
configurations.testImplementation.resolutionStrategy.dependencySubstitution {
	substitute(module(group: 'foo', name: 'bar, version: '1.0', classifier: 'test-helpers', ext: 'jar'))
			.with project(path: ':external:foo:bar', configuration: 'test-helpers')
}

// the -test-helpers.jar artifact is built from the compiled test classes filtering out (*Test)
```

Currently I can achieve this with convoluted dependency setup and an extra module:
 * `:bar`: `main` source set contains production code
   * `testImplementation "foo:bar-test_helpers:1.0"` (this should be `foo:bar:1.0:test-helpers`)
   * \+ a `dependencySubstitution` to replace this with a `project()`
 * `:bar:test_helpers`: extra module that contains test helpers only
   * `compileOnly "foo:bar:1.0"` (did you notice the circle?)
 * `settings.gradle`: have to rename `:bar:test_helpers` at include time  
because otherwise I end up with multiple `test_helpers` projects

(Thanks @oehme for [pointing](https://discuss.gradle.org/t/dependency-substitution-of-modules-with-a-classifier/13518/3) here)

**Comment by ggabdol** [comment]
I'm trying to substitute a module with the same module but with the different classifier, and I don't think the workaround above works for my case.
`substitute module(group: 'foo', name: 'bar', version: '1.0') with module(group: 'foo', name: 'bar', version: '1.0', classifier: 'debug', ext: 'aar')`
Is there any workaround for my case?

**Comment by dryganets** [comment]
The feature looks very promising! 

I was going to use it to modularize the android project, but all my libraries have `aar` extension, not `jar`. Also, I'm using flavors for the libraries and the only viable way to publish such lib to maven is to use different `classifier` for every `productFlavor`.

**Comment by chrisr3** [comment]
We are trying to import a "core" artifact, which has a transitive dependency on
```
co.paralleluniverse:quasar-core:0.7.12_r3:jdk8
```
However, one of the (Java 11) projects which depends on this "core" artifact (in _many_ places) needs to use this Quasar artifact instead:
```
co.paralleluniverse:quasar-core:0.8.0_r3
```
And then Gradle fails, complaining that `quasar-core-0.8.0_r3-jdk8.jar` does not exist.

**Comment by jjohannes** [comment]
@chrisr3 have you tried to register a component metadata rule that 'fixes' the dependency declaration in the _"core" artifacts_? 
https://docs.gradle.org/current/userguide/component_metadata_rules.html#fixing_wrong_dependency_details

Something like:
```
allVariants {
    withDependencies {
        def quasarCoreDep = find { it.name == "quasar-core" }
        if (quasarCoreDep) {
            remove(quasarCoreDep)
            add(group: quasarCoreDep. group, name: quasarCoreDep.name)
        }
    }
}
```

**Comment by chrisr3** [comment]
@jjohannes I have just tried, but it doesn't seem to work in Gradle 5.4.1. Nor can we upgrade to Gradle 6.x _because of Quasar_.

**Comment by jjohannes** [comment]
@chrisr3 although not documented in the user manual of Gradle 5, the dependency modification with rules should work with Gradle 5.x.

The motivation behind component metadata rules is to provide a more fine grained API so that you can adjust the place where something is "wrong", instead of just globally forcing a certain version -which could hide issues if dependencies are changed in the future.

Although I understand that certain use cases are easier to solve with `resolutionStrategy.force`, we concentrated on evolving the component metadata API as we believe it ultimately is safer to use. That's why it now offers more features than the "resolutionStrategy" APIs.

Which error are you seeing?
Did you apply the rule to the right module(s) - the ones you called "core"?

The complete rule applied inline should look somehow like this:
```
dependencies {
  components {
     withModule("your:core-module") {
       allVariants {
         withDependencies {
           def quasarCoreDep = find { it.name == "quasar-core" }
           if (quasarCoreDep) {
             remove(quasarCoreDep)
             add(group: quasarCoreDep.group, name: quasarCoreDep.name)
           }
         }
      }
    }
  }
}
```

Or you put the rule into a class as shown in the docs and then you can do:

```
dependencies {
  components {
     withModule("your:core-module", CoreModuleRule)
     withModule("your:core-module2", CoreModuleRule)
     withModule("your:core-module3", CoreModuleRule)
     ...
  }
}
```

**Comment by mauromol** [comment]
If we could replace a module with the same module but different classifier we would probably be able to easily force Groovy dependency to be `indy` (see #9318), wouldn't we?

**Comment by melix** [comment]
It would actually be easier to model the "indy" version of Groovy as a variant. Then you don't have to specify that you need a specific classifier, the consumer could choose whether to use indy or not.

**Comment by mauromol** [comment]
> It would actually be easier to model the "indy" version of Groovy as a variant. Then you don't have to specify that you need a specific classifier, the consumer could choose whether to use indy or not.

If I understand you right, you're talking about a change in Groovy published metadata. My previous message and #9318 was about making it easier for consumers to use Groovy as it is right now. It has become harder to consume indy Groovy now that it has been split into multiple modules, with dependencies between each other (as I reported in #9318).

**Comment by melix** [comment]
Ideally done in Groovy published metadata, but this can also be done via component metadata rules on Gradle side. This is how, for example, we solve the guava-jre vs guava-android mess.

**Comment by mauromol** [comment]
So you mean this: https://docs.gradle.org/current/userguide/component_metadata_rules.html#making_variants_encoded_in_versions_explicit
Do you really think this is easier from a Gradle user point of view rather than something like the following (if it worked)?

```groovy
configurations.all {
	resolutionStrategy.dependencySubstitution {
		substitute module('org.codehaus.groovy:groovy:2.5.10') with module('org.codehaus.groovy:groovy:2.5.10:indy')
	}
}
```

Honestly, I refuse to put all that code required for component metadata rules in my build script... I rather prefer to use exclusions, even if philosophically less correct...

**Comment by melix** [comment]
First, yes I think it's better. It's better because it's not just about the groovy runtime. This is also about all libraries compiled with the "indy" runtime. This is missing information today, and you _will_ run into problems if you mix indy and non indy dependencies.

Second, about bloating the build script: I fully agree with you, this is why this should be the responsibility of a plugin to do this, so that you don't have to write it: all users like you would just have to apply the same plugin. This is not any different from the logger plugin that @ljacomet wrote, or the plugin that @jjohannes wrote for Guava. 

Of course, it's better if  Groovy and the Groovy team provide standard attributes to do this. Now I am unsure they'd be willing to do so, given that Groovy 4 might drop the legacy runtime. But consistency is better than everybody patching, or you don't make the ecosystem better: you're just making everybody's life harder.

**Comment by chrisr3** [comment]
I'm with @mauromol on this one. I have a similar problem where the _glaringly obvious_ solution of
```
configurations.all {
    resolutionStrategy.dependencySubstitution {
        substitute module('<my-group>:<my-artifact>') with module('<my-group>:<my-artifact>:<new-version>')
    }
}
```
_**DOES NOT WORK**_ because the original dependency has a classifier, and the substitution mechanism is incapable of removing/overriding it.

Gradle's current behaviour is completely counter-intuitive - it is effectively trying to resolve a unique Maven artifact using only 80% of that artifact's primary key, viz its Maven coordinate. (And I say that _on the assumption_ that Maven's `packaging` coordinate is fully supported here...) That you then present the `ComponentMetadataRule` API as a "solution" to this design flaw is just jaw-dropping.

**Comment by ahmed3elshaer** [comment]
I'm also trying to do the same as you guys, I have a dependency that is implemented like this 
```
        productionImplementation group: "$staticPackage", name: 'module', version: "$dynamicReleaseVersion", ext: "$extension", classifier: "$prod"
```
and i want to use 
```
 resolutionStrategy.dependencySubstitution {
            substitute module( group: "$staticPackage", name: 'module', version: "$dynamicReleaseVersion", ext: "$extension", classifier: "$prod") with project(':local')
}
```
Does anyone managed to do it ?

**Comment by chrisr3** [comment]
> Does anyone managed to do it ?

We decided that the cleanest solution was to fork the other project to remove its problematic classifier at source. _**Seriously.**_

**Comment by ahmed3elshaer** [comment]
> > Does anyone managed to do it ?
> 
> We decided that the cleanest solution was to fork the other project to remove its problematic classifier at source. _**Seriously.**_

I did a work around and changed my dependency to different artifacts and replace the job of the   `extension` & `classifier` with the artifactId as a unique identifier and it seems to be working

**Comment by melix** [comment]
_I would expect some comments here to be a bit nicer and follow our [code of conduct](https://gradle.org/conduct/)._ 

I think there's a confusion on what we _want_ to do here. There are a couple of different issues:

1. what producers should do. Libraries like Groovy, Guava, etc... should avoid using classifiers as much as possible, and rather use what we provide, richer _variant declaration_, which avoids those problems altogether from day 1.
2. if a producer, like most today, doesn't do that, it's possible with Gradle to enrich existing metadata to make it possible. The advantage of doing so is that again the consumer doesn't have to say anything, except "give me the indy version everywhere" or "give me the jre version of Guava". Actually in most cases you wouldn't say _anything_ and it would work. This logic should be bundled in plugins like what the logging plugin does, so that you don't have to repeat in your own builds.

And then there's the last issue, which is _convenience_ for existing Maven metadata out there, a possibility to perform dependency substitution based on the GAV coordinates + classifier (type and extension is different as they are _artifact_ selectors). It would solve most of the problems listed here, although this cannot be seen as _the_ solution: it's a convenience, but it doesn't fix the root cause.

**Comment by saudet** [comment]
@melix The main problem with variants is that, *by design*, we can't add to the dependency graph more than a single variant of each dependency. That's (one of the main reasons) why we still need classifiers. There is such limitations with classifiers. Whatever the solution may be, variants are not it.

For reference, to get the classifiers of dependencies, I was able to hack something together that works with Gradle 5.0+ and ComponentMetadataRule: https://github.com/bytedeco/gradle-javacpp/blob/master/src/main/java/org/bytedeco/gradle/javacpp/PlatformRule.java

**Comment by melix** [comment]
> @melix The main problem with variants is that, by design, we can't add to the dependency graph more than a single variant of each dependency. That's (one of the main reasons) why we still need classifiers. There is no such limitations with classifiers. Whatever the solution may be, variants are not it.

They are mutually exclusive _if they provide the same capability_. The capability concept lets you declare that two variants _can_ be found in a single graph. I'm not saying that this is always the solution, it all depends on the use case, but the variant model _does_ handle that.

Having said that, there's a PR ready to make this "classifier"/old world easier to deal with: https://github.com/gradle/gradle/pull/13142

**Comment by saudet** [comment]
> They are mutually exclusive _if they provide the same capability_. The capability concept lets you declare that two variants _can_ be found in a single graph. I'm not saying that this is always the solution, it all depends on the use case, but the variant model _does_ handle that.

Correct me if I'm wrong, but a capability is just a GAV, there's no classifier or anything. I fail to see how that makes a difference.

> Having said that, there's a PR ready to make this "classifier"/old world easier to deal with: #13142

Awesome! Please also consider fixing snapshots with classifiers (issue #2882).

**Comment by melix** [comment]
> Correct me if I'm wrong, but a capability is just a GAV, there's no classifier or anything. I fail to see how that makes a difference.

The Gradle model has variants. A variant is attached with a number of artifacts. A "classifier" is just an artifact. So the right way to model things is to have different variants and attach the artifacts, classified or not, to the right variants. A single component can provide different variants with different capabilities (for example, that's what we do for test fixtures), in which case they are allowed to be on the same dependency graph. For example, you can have the "test fixtures" variant, which carries the `foo-1.0-test-fixtures.jar` classified artifact, along with the `foo-1.0.jar` main artifact on a single graph.

**Comment by saudet** [comment]
> The Gradle model has variants. A variant is attached with a number of artifacts. A "classifier" is just an artifact. So the right way to model things is to have different variants and attach the artifacts, classified or not, to the right variants. A single component can provide different variants with different capabilities (for example, that's what we do for test fixtures), in which case they are allowed to be on the same dependency graph. For example, you can have the "test fixtures" variant, which carries the `foo-1.0-test-fixtures.jar` classified artifact, along with the `foo-1.0.jar` main artifact on a single graph.

How about the javadoc and source artifacts? Why are those not modeled and published as variants?
https://docs.gradle.org/current/userguide/publishing_maven.html

**Comment by melix** [comment]
They are:

```
java {
    withJavadocJar()
    withSourcesJar()
}
```

configures the tasks and publishes them as classified artifacts (for Maven and other tools to use them) but are properly modelled as variants in Gradle Module Metadata.

**Comment by saudet** [comment]
> They are:
> 
> ```
> java {
>     withJavadocJar()
>     withSourcesJar()
> }
> ```
> 
> configures the tasks and publishes them as classified artifacts (for Maven and other tools to use them) but are properly modelled as variants in Gradle Module Metadata.

Ah, they are, ok. Let me rephrase that, why use classifiers? Why not append like `-javadoc` and `-sources` to the artifact name and publish them with separate unrelated pom.xml files? I'm guessing the answer is in your reply above: Because other tools rely on them! In my opinion, that's a perfectly valid reason to actually support classifiers properly. If you're serious about the following claim, classifiers absolutely need to be supported properly:

> In fact, Gradle works perfectly with Maven-compatible repositories, which makes it easy to migrate your dependencies.

https://docs.gradle.org/current/userguide/migrating_from_maven.html

**Comment by melix** [comment]
Let me reformulate. Gradle's model doesn't care what file name you use. It cares about metadata and variants. A variant tells which files it needs, which is infinitely superior to the poor "classifier" approach which has no semantic whatsoever associated with it.

There are, however, classified artifacts everywhere in the wild. In some cases, they should come with their own dependencies. In others, they shouldn't have any dependency (like javadocs, sources, ...). The classifier doesn't tell anything about _why_ it's there. Is it because it's another alternative for a different JDK (hello, `-jdk8`), or it it because it's a fat jar (hello, `-all`), or is it a sources jar? It's completely _wild_.

Gradle variants, on the contrary, define what they are used for, and attach _artifacts_ to them. The relevant part is the metadata, not the artifact.

Now, _because_ there are lots of tools in the wild, _because_ there's also Maven, it's _convenient_ that we publish some of those variant artifacts _using a classifier_, because it allows some behavior for those tools.

In the "javadoc" example, it's pretty simple, the additional variant we publish has the classifier, and the variant says it has no dependencies. Because Maven and IDEs will only download the javadoc jar, that's easy.

In other cases, like "test fixtures", the Gradle model is _way better_. Because it allows us to tell "if you pick this test fixtures variant, then you need those additional dependencies which are _only_ required for testing". We publish them as _classified_ artifacts, because it allows _Maven consumers_ to pick them. They would _not_ pick the dependencies we're able to give if the consumer is Gradle, but they _can_ use it. In other words, the Gradle model allows us to be better when the consumer is a Gradle consumer, while preserving _some_ features for Maven.

If you think the other way around, and publish classifiers _without_ Gradle variants (that is to say without associated Gradle metadata), you're starting to face **exactly** what this whole thread is about. Because there's no associated semantics, Gradle _cannot_ know what to do. For Maven it's even worse because the only solution you have is to exclude dependencies and reinclude new ones, even if you don't directly use them!

Then, there's the "Maven-compatible repositories". Gradle **is** compatible with Maven repositories. Even if you use Gradle metadata, we **do** publish and consume from Maven repositories (and Ivy repositories too). The fact that those repositories contain classified javadocs is _also_ supported. In fact, it's _because_ we support them that you're facing this issue.

So, again that's why we think that the _right model_ is, when you have a library that is published with classifiers, is to actually _explain why they are there_, which in Gradle terms turns into "describe them as variants". If you are the producer and can support Gradle metadata, it's a win-win, everybody is happy. If you can't, then there are two options:

1. use component metadata rules to "fix" this missing metadata
2. use the PR mentioned above when you don't have time to think about it, or want a quick fix.

**Comment by saudet** [comment]
I'm not arguing about Gradle's model at all. It might be the best thing since sliced bread! I honestly don't know. I'm sorry if I gave the wrong impression. Think of it that way: If Gradle allowed people to migrate to that model easily, but still maintain compatibility with other projects, teams, etc, then it would make things a lot easier.

> Then, there's the "Maven-compatible repositories". Gradle **is** compatible with Maven repositories. Even if you use Gradle metadata, we **do** publish and consume from Maven repositories (and Ivy repositories too). The fact that those repositories contain classified javadocs is _also_ supported. In fact, it's _because_ we support them that you're facing this issue.

That is untrue. Gradle doesn't correctly support, for example and as mentioned above, snapshots with classifiers. @carlspring opened issue #2882 about that a few years ago, and it was basically closed as "won't fix". Why is that? If Maven compatibility isn't a priority, that's fine, but at least please update your documentation about it.

**Comment by melix** [comment]
Please don't pollute this thread with an unrelated discussion. The conversation about #2882 is irrelevant to this issue and everything is actually explained in the comments of that ticket.

**Comment by ljacomet** [comment]
Improvements have been made and will be released with Gradle 6.6

**Comment by chrisr3** [comment]
> Improvements have been made and will be released with Gradle 6.6

This doesn't help in the case of Quasar: Quasar is incompatible with Kotlin 1.3 byte-code, but the Kotlin 1.2.x Gradle plugins are incompatible with Gradle 6.x+. (They use an API that was subsequently removed.) Could these improvements be back-ported to a Gradle 5.6.5 release please?

**Comment by TWiStErRob** [comment]
@chrisr3 that's not possible, since if 5.6.5 includes it, then 6.0 has to as well, but it's 6.6 that has it.

**Comment by ljacomet** [comment]
There will be no backport, especially as a lot of things happened in Gradle 6 around dependency management which could make the backporting way too large.

**Comment by chrisr3** [comment]
I didn't expect any action really, but Gradle 6.x _is_ still dead to us for the reason given.
